### DISCUSSION TITLE
Is possible to create a snippet that when expanded it insert a line in the start of the file?

### DISCUSSION THREAD

**Discussion by LaBatata101** [L3MON4D3/LuaSnip#383] [Q&A] [answered]
I'm trying to create a snippet like this:
```
from dataclasses import dataclass

@dataclass
class {1}:
    {0}
```
When I type `dataclass` it would expand the snippet and insert the line `from dataclasses import dataclass` at the start of the file.
But I have no idea how to do this. Can someone help me?

**Comment by L3MON4D3** [comment, marked as answer]
Ouh, yeah that's possible: just create a functionNode without arg nodes. The code in there will only run once during the initial expansion and you won't see it as long as it ~doesn't return any text~ returns an empty string. 
```lua
s("dataclass", {
    <other nodes>, 
    f(insert_dataclass_at_beginning, {})
})
```
This will be more straightforward once pre_expand/post_expand-callbacks are added.

**Reply by LaBatata101** [reply]
Cool! So I came up with this code:
```lua
ls.add_snippets("python", {
  s("dataclass", {
    t({ "@dataclass", "class " }),
    i(1),
    t({ ":", "\t" }),
    i(0),
    f(function(_, _)
      vim.api.nvim_buf_set_lines(0, 0, 0, true, { "from dataclasses import dataclass" })
      return ""
    end, {}),
  }),
})
```
But I'm getting this error when I try to expand the snippet:
```
E5108: Error executing lua ...e/nvim/site/pack/packer/opt/LuaSnip/lua/luasnip/init.lua:179: end_col out of bounds
stack traceback:
        [C]: in function 'nvim_buf_set_text'
        ...e/nvim/site/pack/packer/opt/LuaSnip/lua/luasnip/init.lua:179: in function 'snip_expand'
        ...e/nvim/site/pack/packer/opt/LuaSnip/lua/luasnip/init.lua:242: in function 'expand'
        ...e/nvim/site/pack/packer/opt/LuaSnip/lua/luasnip/init.lua:289: in function 'expand_or_jump'
        /home/labatata/.config/nvim/lua/config/plugins/cmp.lua:28: in function 'on_keymap'
        ...hare/nvim/site/pack/packer/opt/nvim-cmp/lua/cmp/core.lua:145: in function 'callback'
        ...m/site/pack/packer/opt/nvim-cmp/lua/cmp/utils/keymap.lua:114: in function <...m/site/pack/packer/opt/nvim-cmp/lua/cmp/utils/keymap.lua:108>
        ...m/site/pack/packer/opt/nvim-cmp/lua/cmp/utils/keymap.lua:244: in function <...m/site/pack/packer/opt/nvim-cmp/lua/cmp/utils/keymap.lua:243>
```
And it gets expanded wrong, the trigger text it's not getting removed:
```
dataclass@dataclass
class :
```
Not sure what I'm doing wrong.

**Reply by L3MON4D3** [reply]
Ofc🤦‍♂️: we store the position the trigger was at and clear it only after the snippet was expanded, if some text is inserted during expansion, it'll fail.
You can `vim.schedule` the insertion, that way it should work.

**Reply by LaBatata101** [reply]
Thanks, that worked!

**Comment by leiserfg** [comment]
Functions are  evaluated withing the previews, I think better to use some kind of guard to avoid rewriting the line.

**Reply by L3MON4D3** [reply]
Oh right, thats also important, otherwise you'll get the text even when scrolling over the snippet via `nvim-cmp` (if documentation is enabled at least)

**Reply by LaBatata101** [reply]
Thanks for the info!
